app/src/main/java/org/thoughtcrime/securesms/migrations/ApplicationMigrations.java

```diff
@@ -0,0 +1,398 @@
+      }
+
+      if (uiBlockingVersion > lastSeenVersion) {
+        Log.i(TAG, "Migration set is UI-blocking through version " + uiBlockingVersion + ".");
+        UI_BLOCKING_MIGRATION_RUNNING.setValue(true);
+      } else {
+        Log.i(TAG, "Migration set is non-UI-blocking.");
+        UI_BLOCKING_MIGRATION_RUNNING.setValue(false);
+      }
+
+      final long startTime = System.currentTimeMillis();
+      final int  uiVersion = uiBlockingVersion;
+
+      EventBus.getDefault().register(new Object() {
+        @Subscribe(sticky = true, threadMode = ThreadMode.MAIN)
+        public void onMigrationComplete(MigrationCompleteEvent event) {
+          Log.i(TAG, "Received MigrationCompleteEvent for version " + event.getVersion() + ". (Current: " + CURRENT_VERSION + ")");
+
+          if (event.getVersion() > CURRENT_VERSION) {
+            throw new AssertionError("Received a higher version than the current version? App downgrades are not supported. (received: " + event.getVersion() + ", current: " + CURRENT_VERSION + ")");
+          }
+
+          Log.i(TAG, "Updating last migration version to " + event.getVersion());
+          TextSecurePreferences.setAppMigrationVersion(context, event.getVersion());
+
+          if (event.getVersion() == CURRENT_VERSION) {
+            Log.i(TAG, "Migration complete. Took " + (System.currentTimeMillis() - startTime) + " ms.");
+            EventBus.getDefault().unregister(this);
+
+            VersionTracker.updateLastSeenVersion(context);
+            UI_BLOCKING_MIGRATION_RUNNING.setValue(false);
+          } else if (event.getVersion() >= uiVersion) {
+            Log.i(TAG, "Version is >= the UI-blocking version. Posting 'false'.");
+            UI_BLOCKING_MIGRATION_RUNNING.setValue(false);
+          }
+        }
+      });
+    } else {
+      Log.d(TAG, "No migrations.");
+      TextSecurePreferences.setAppMigrationVersion(context, CURRENT_VERSION);
+      VersionTracker.updateLastSeenVersion(context);
+      UI_BLOCKING_MIGRATION_RUNNING.setValue(false);
+    }
+  }
+
+  /**
+   * @return A {@link LiveData} object that will update with whether or not a UI blocking migration
+   * is in progress.
+   */
+  public static LiveData<Boolean> getUiBlockingMigrationStatus() {
+    return UI_BLOCKING_MIGRATION_RUNNING;
+  }
+
+  /**
+   * @return True if a UI blocking migration is running.
+   */
+  public static boolean isUiBlockingMigrationRunning() {
+    Boolean value = UI_BLOCKING_MIGRATION_RUNNING.getValue();
+    return value != null && value;
+  }
+
+  /**
+   * @return Whether or not we're in the middle of an update, as determined by the last seen and
+   * current version.
+   */
+  public static boolean isUpdate(@NonNull Context context) {
+    return isLegacyUpdate(context) || TextSecurePreferences.getAppMigrationVersion(context) < CURRENT_VERSION;
+  }
+
+  private static LinkedHashMap<Integer, MigrationJob> getMigrationJobs(@NonNull Context context, int lastSeenVersion) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 44. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.